Reallife training tests

LeonidBeynenson opened the pull request:
Add one reallife training test: 
* model gen3_mobilenetV2_ATSS, 
* dataset bbcd

LeonidBeynenson committed:
Chnages in patience-s in gen3_mobilenetV2_ATSS config for reallife training tests

These are the latest changes from gzalessk.

LeonidBeynenson committed:
Add one BBCD reallife training test (without check result quality yet)

Also add new field constants DEFAULT_FIELD_VALUE_FOR_USING_IN_TEST and
KEEP_FIELD_VALUE_FROM_CONFIG.

LeonidBeynenson committed:
Add validation of reallife tests results in test_ote_training.py

LeonidBeynenson committed:
Merge branch 'lb/add_template_for_pot_nncf_tests' into lb/reallife_training_tests

LeonidBeynenson committed:
Add comments and docstrings to new fixtures in test_ote_training.py

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/reallife_training_tests

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/reallife_training_tests

druzhkov-paul commented on tests/expected_metrics/metrics_test_ote_training.yml:
```diff
@@ -0,0 +1,20 @@
+                'max_drop': 0.01
+'ACTION-pot_evaluation,model-gen3_mobilenetV2_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife':
+        'metrics.accuracy.f-measure':
+                'base': 'training_evaluation.metrics.accuracy.f-measure'
+                'max_drop': 0.01
```
Not sure if this what others think of referring to the accuracy drop during compression. @alexsu52 Could you please comment on this?

LeonidBeynenson replied:
Usually max accuracy drop is 1% w.r.t. the original training accuracy.
@alexsu52, could you comment it?

alexsu52 replied:
Of course, the customer is interested in the model drop value relative to the Pytorch FP32 model. But this drop includes the export drop value + the compression drop value. Thus you can specify the drop in the 1% w.r.t. the original training accuracy if the model is exported without loss of accuracy. If you can't guarantee this you should calculate the drop relative to IR FP32 model.

LeonidBeynenson replied:
>  If you can't guarantee this you should calculate the drop relative to IR FP32 model.

@alexsu52 : Ok, Alexander, thank you.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -111,6 +138,94 @@ def template_paths_fx(request):
+    The keys of the dict are the parameters' part of the test id-s -- see the function
+    TestOTEIntegration._generate_test_id.
+    The value for each key is a structure that stores a requirement on some metric.
+    The requirement can be either a target value (probably, with max size of quality drop)
```
What is max size? Size of what?

LeonidBeynenson replied:
Sorry, Pavel, do not understand your question. 
Max size of quality drop -- usually it is 1%, e.t. 0.01.
Do you have suggestion how to write it more clear?

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -111,6 +138,94 @@ def template_paths_fx(request):
+    """
+    This fixture returns the expected metrics for the current test.
+    It may be
+    * either a dict with the structure that stores the requirements on
```
"either" means that there're two options. Not three.

LeonidBeynenson replied:
Ok

druzhkov-paul commented on tests/test_ote_training.py:
````diff
@@ -111,6 +138,94 @@ def template_paths_fx(request):
+      }
+      ```
+    * or a string with the special value `ABSENT_EXPECTED_VALIDATION_METRICS`
+      if the structure that stores the requirements on metrics 3for the current
````
```suggestion
      if the structure that stores the requirements on metrics for the current
```

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -243,15 +363,24 @@ def _run_ote_training(self, data_collector):
+            logger.debug(f'Keep params.learning_parameters.num_iters={params.learning_parameters.num_iters}')
+
+        if params.learning_parameters.num_iters < 20:
             num_checkpoints = 2
-        elif self.num_training_iters < 1000:
+        elif params.learning_parameters.num_iters < 1000:
             num_checkpoints = 10
         else:
             num_checkpoints = 30
```
I don't think, that we need this logic anymore. `num_checkpoints` is not used in the implementation now.

LeonidBeynenson replied:
But will this value be used in the future?
If you are sure that in the nearest future it won't be used, I will remove this block of code.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -332,8 +463,14 @@ def __call__(self, data_collector: DataCollector,
+        score_name, score_value = self._run_ote_evaluation(data_collector, **kwargs)
+        results = {
+                'metrics': {
+                    'accuracy': {
```
What else except accuracy can be here?

LeonidBeynenson replied:
Time of execution of a stage (e.g. time of POT, etc)
It is planned.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -643,10 +806,80 @@ def __call__(self, data_collector: DataCollector,
+            raise RuntimeError(f'Wrong expected metrics {current_test_expected_metrics}')
+
+        if 'max_drop' in cur_metric_requirements:
+            target_value = target_value - float(cur_metric_requirements['max_drop'])
```
I'd suggest comparing the absolute difference.

LeonidBeynenson replied:
No, usually the more is the better.
Note that when we measure time of stage execution in the future (that is "the less the better"), we will require to point direction "what is better -- more or less" in the YAML file (as in accuracy checker).
But I'd like to finish this PR as is now.

druzhkov-paul replied:
It might be the common way to compare the difference for NNCF/POT compression. Let's @alexsu52 @AlexanderDokuchaev better comment on this. But for ONNX/OpenVINO export we are not expecting the metric to become higher that it was for original model. If, given a limited amount of data, we end up seeing much larger metric values for the exported model then most likely we do have a bug.

alexsu52 replied:
I agree with @druzhkov-paul. This statement is valid for the POT and for the NNCF export and validation tasks. We can face the issue with NNCF because we can stop optimization only at the end of an epoch training. My general proposal is to use `abs` for all tasks except the NNCF optimization task. Probably, the comparator type needs to add to the config.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -643,10 +806,80 @@ def __call__(self, data_collector: DataCollector,
+    logger.info(f'Validation: end, result={is_passed}')
+    if not is_passed:
+        fail_reasons = '\n'.join(fail_reasons)
+        pytest.fail(f'Validation is failed:\n{fail_reasons}')
```
```suggestion
        pytest.fail(f'Validation failed:\n{fail_reasons}')
```

druzhkov-paul reviewed (commented):
@LeonidBeynenson I want to clarify the overall flow here. Please correct me if my understanding is incorrect:

- All tests stages are run only once.
- If some stage is subject to be verified w.r.t produced quality, then it will happen:
  - Only once.
  - During the call to a respective stage, not the one that depends on it.
- If quality evaluation fail then corresponding test will fail, but all dependent tests will be run.
-

LeonidBeynenson committed:
Small update test_ote_training.py

Co-authored-by: Pavel Druzhkov <druzhkov.paul@gmail.com>

LeonidBeynenson committed:
Remove using `num_checkpoints` parameter from test_ote_training.py

LeonidBeynenson committed:
Decorative change in test_ote_training.py

Co-authored-by: Pavel Druzhkov <druzhkov.paul@gmail.com>

alexsu52 commented on tests/expected_metrics/metrics_test_ote_training.yml:
```diff
@@ -0,0 +1,20 @@
+'ACTION-export_evaluation,model-gen3_mobilenetV2_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife':
+        'metrics.accuracy.f-measure':
+                'base': 'training_evaluation.metrics.accuracy.f-measure'
+                'max_drop': 0.01
```
I would suggest using a stricter constraint for the "export evaluation" case. I propose using a minimally available value, for example, 0.001 or less.

druzhkov-paul replied:
Not sure if 0.1% is reachable here at all. 1% seems reasonable for a default value.

alexsu52 commented on tests/expected_metrics/metrics_test_ote_training.yml:
```diff
@@ -0,0 +1,20 @@
+'ACTION-nncf_export_evaluation,model-gen3_mobilenetV2_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife':
+        'metrics.accuracy.f-measure':
+                'base': 'training_evaluation.metrics.accuracy.f-measure'
+                'max_drop': 0.01
```
I would recommend specifying two drop values and both criteria must be passed:
- 'base': 'training_evaluation.metrics.accuracy.f-measure'
  'max_drop': 0.01
- 'base': 'nncf_evaluation.metrics.accuracy.f-measure'
  'max_drop': 0.003

LeonidBeynenson committed:
Add validation step to test_ote_training.py for stages that were run from dependency chain earlier

It is possible that a stage is run from a dependency chain (for some
other stage) and later it is run not from dependency chain.
In this case the stage's action will be run in the first run only, but
validation will be run each time when the stage is run not from a
dependency chain.

LeonidBeynenson commented:
> @LeonidBeynenson I want to clarify the overall flow here. Please correct me if my understanding is incorrect:
>
> * All tests stages are run only once.
> * If some stage is subject to be verified w.r.t produced quality, then it will happen:
>   * Only once.
>   * During the call to a respective stage, not the one that depends on it.
> * If quality evaluation fail then corresponding test will fail, but all dependent tests will be run.

Yes, as far as I can see your understanding is absolutely correct.

Some additional details:
* Each action will be run not more than once, the result of the action will be re-used.
  Note that if you point a concrete stage as a pytest argument, then all the dependency of this stage are run first, then this pointed stage is run.
  When dependent stages are called, they can call their own dependent stages -- they will compose so called "dependency chain"
* If a stage should be verified, then verification is done if and only if we call this stage not from a dependency chain.
* Theoretically it is possible that a stage is run from a dependency chain (for some other stage) and later it is run not from dependency chain. 
  In this case the stage's action will be run in the first run only, but validation will be run each time when the stage is run not from a dependency chain 
  Pavel, please see my new changes in the commit 5ebb77ce -- it improves behavior in this case.

* And yes, you are right, if quality evaluation fails, the test is failed too, but its results may be re-used in later tests.

LeonidBeynenson committed:
Move getting expecting metric to callback (in test_ote_training.py)

It is required to avoid attempts to get expected validation metrics for
tests that do not require validation at all.

LeonidBeynenson committed:
Give possibility to point complicated acceptable ranges of expected metrics in test_ote_training.py

LeonidBeynenson commented:
2 @druzhkov-paul , @alexsu52 :
Made couple of changes:
*  Moved getting expecting metric to callback
    It was required to avoid attempts to get expected validation metrics for
    tests that do not require validation at all.
    See the commit bd90e886
* Gave possibility to point complicated acceptable ranges of expected metrics 
    Now ranges and half-bounded ranges may be used : you can set in expected metrics either `max_diff` (for symmetric range) or one of `max_diff_if_less_threshold` and `max_diff_if_greater_threshold` (for half-bounded range), or both `max_diff_if_less_threshold` and `max_diff_if_greater_threshold` (for asymmetric range)
    See the commit c5954b1a.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -111,6 +128,112 @@ def template_paths_fx(request):
+
+@pytest.fixture
+def cur_test_expected_metrics_callback_fx(expected_metrics_all_tests_fx, current_test_parameters_string_fx,
+                                          current_test_parameters_fx) -> Union[None, Callable[[],Dict]]:
```
nit:
`Union[None, X]` == `Optional[X]`.

druzhkov-paul commented on tests/expected_metrics/metrics_test_ote_training.yml:
```diff
@@ -0,0 +1,21 @@
+'ACTION-training_evaluation,model-gen3_mobilenetV2_ATSS,dataset-bbcd,num_iters-KEEP_CONFIG_FIELD_VALUE,batch-KEEP_CONFIG_FIELD_VALUE,usecase-reallife':
+        'metrics.accuracy.f-measure':
+                'target_value': 0.81
+                'max_diff_if_less_threshold': 0.005
+                'max_diff_if_greater_threshold': 0.015
```
What does that mean?

druzhkov-paul replied:
My guess is that this is equivalent to:
```
-max_diff_if_less_threshold < actual - expected < max_diff_if_greater_threshold
```
@LeonidBeynenson, right?

LeonidBeynenson replied:
Sorry that did not answer.
Yes, you are right: if both of the values `max_diff_if_less_threshold` and `max_diff_if_greater_threshold` are set, then it works as you wrote.
But we can point only one of the values -- in this case the range of accessible values will be half-bounded.

druzhkov-paul replied:
I just wanted to emphasize, that IMO naming is not perfect, at least.

LeonidBeynenson replied:
Ohh, Pavel, I agree with you.
If you have ideas how to call them, please, propose your version!

alexsu52 reviewed (approved):
Looks good to me

LeonidBeynenson committed:
Make evaluation on TESTING subset in test_ote_training.py

LeonidBeynenson committed:
Add one more exception handling in test_ote_training.py

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/reallife_training_tests

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/reallife_training_tests

LeonidBeynenson committed:
Improve handling wrong config errors in test_ote_training.py

Also add comments and todos.

LeonidBeynenson commented:
Hi @druzhkov-paul ,   
I made several checks how works the code that makes 
* parsing the test expected metrics, 
* comparison of result test metrics w.r.t. the test expected metrics, 
* and validation
and fixed some small issues in the code.
(You can see the list of checks in the file [diff_test_ote_training_as_test_report.md](https://github.com/openvinotoolkit/mmdetection/files/7576993/diff_test_ote_training_as_test_report.md) )

I think now this PR is ready for merge.

LeonidBeynenson commented:
Integration tests are passed as CI job lbeynens-596-test.
Reallife tests are started as CI job PERF-lbeynens-610-test

LeonidBeynenson commented:
CI job PERF-lbeynens-610-test was failed, I re-ran it, after that  PERF-lbeynens-612-test was passed
